libp2p/Host.cpp

```diff
@@ -127,16 +132,21 @@ void Host::stop()
     // such tasks may involve socket reads from Capabilities that maintain references
     // to resources we're about to free.
 
-    // ignore if already stopped/stopping, at the same time,
-    // signal run() to prepare for shutdown and reset m_timer
-    if (!m_run.exchange(false))
+    // ignore if already stopped/stopping, at the same time
+    // indicates that the network is shutting down
+    if (!haveNetwork())
         return;
+    
+    if (haveCapabilities())
```

Comment:
Do we need this check? Can't we make `m_run` have sensible values in both modes?

Reply:
Good point - I've reworked my changes so that m_run is now always set to true if the network thread has been started, I've added a `haveCapabilities()` call where appropriate.